### PR TITLE
﻿Catch the mono_class_from_name exception and improve exception logging

### PR DESCRIPTION
**Pull request by alexey-zakharov**
Some Unity Projects fail initialization due to TypeLoadException. Example could be [2D Performance PlayMode Tests project](https://github.cds.internal.unity3d.com/unity/neutron/tree/main/Tests/PerformanceTests/2D) with the [type visibility reason](https://unity.slack.com/archives/C02LE0UP341/p1685131806842419).

As `mono_class_from_name` is not autogenerated and doesn't have automatic try/catch wrapping in C# implementation  I would like to add some diagnostic to the native part of it to simplify troubleshooting for further projects.

Example of logging:
```
CoreCLR Exception 1380929364: Type 'UnityEngine.U2D.IK.IKManager2D' from assembly 'Unity.2D.IK.Runtime, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' is attempting to implement an inaccessible interface.
```

### CODE REVIEW

**Review by joncham — APPROVED** [review]
I approve, but defer to @mrvoorhe and @UnityAlex as they have been working in this area.

File: src/coreclr/vm/mono/mono_coreclr.cpp [hunk 1]
```diff
@@ -481,9 +481,33 @@ extern "C" EXPORT_API MonoClass* EXPORT_CC mono_class_from_mono_type(MonoType *i
...
+    CONTRACTL_END;
+
     GCX_PREEMP(); // temporary until we sort out our GC thread model
 
-    return g_HostStruct->class_from_name(image, name_space, name, ignoreCase);
+    MonoClass* monoClass = NULL;
+
+    EX_TRY
+    {
+        monoClass = g_HostStruct->class_from_name(image, name_space, name, ignoreCase);
```
Comment (mrvoorhe) [inline src/coreclr/vm/mono/mono_coreclr.cpp:500]: Last week I landed https://github.com/Unity-Technologies/runtime/pull/178 which means this is never going to throw.  If this is never going to throw, then I don't see how adding a try/catch here is going to do anything.

File: src/coreclr/vm/mono/mono_coreclr.cpp [hunk 2]
```diff
@@ -481,9 +481,33 @@ extern "C" EXPORT_API MonoClass* EXPORT_CC mono_class_from_mono_type(MonoType *i
...
+    }
+    EX_CATCH
+    {
+        SString sstr;
+        GET_EXCEPTION()->GetMessage(sstr);
+        unity_log("CoreCLR Exception %d: %s\n", GET_EXCEPTION()->GetType(), sstr.GetUTF8());
+    }
+    EX_END_CATCH(SwallowAllExceptions)
```
Comment (mrvoorhe) [inline src/coreclr/vm/mono/mono_coreclr.cpp:508]: If this catch was hit, it's at odds with the stance we decided with https://github.com/Unity-Technologies/runtime/pull/178  that we would make exceptions fatal.

Do we want to establish fatal vs non-fatal exceptions?

File: src/coreclr/vm/mono/mono_coreclr.cpp [hunk 3]
```diff
@@ -481,9 +481,33 @@ extern "C" EXPORT_API MonoClass* EXPORT_CC mono_class_from_mono_type(MonoType *i
 // Note: This function has NOT been generated by the source generator
 MonoClass * mono_class_from_name(MonoImage *image, const char* name_space, const char *name, bool ignoreCase)
 {
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        PRECONDITION(image != NULL);
```
Comment (mrvoorhe) [inline src/coreclr/vm/mono/mono_coreclr.cpp:488]: If the goal is to move the embedding api to C#, then shouldn't we also put argument validation in C#?

File: src/coreclr/vm/mono/mono_coreclr.cpp (hunk 1 above)
Comment (alexey-zakharov) [reply src/coreclr/vm/mono/mono_coreclr.cpp:500]: @mrvoorhe I couldn't test on the latest as #178 is not yet in the unity repo (requires attribute methods changes as DO_API fail), if you say that try/catch is automatically added to `g_HostStruct->class_from_name` then there is no need in the PR :) 
I can see we have it here https://github.com/Unity-Technologies/runtime/pull/178/files#diff-4d3791b94ebacb3f48f99ad8b0cb63cd2a6bd48932f2ecb1122b9b355b6865aeR84, but I wasn't sure whether or not is covers `mono_class_from_name` as well as the comment says it is not generated

File: src/coreclr/vm/mono/mono_coreclr.cpp (hunk 2 above)
Comment (alexey-zakharov) [reply src/coreclr/vm/mono/mono_coreclr.cpp:508]: @mrvoorhe fyi - on the 2D project the exception was not fatal. it caused load failures for objects which had not loaded types, but 80% of tests were passing. IMHO it is best to keep it fatal while we get to the stable coreclr player and editor

File: src/coreclr/vm/mono/mono_coreclr.cpp (hunk 1 above)
Comment (mrvoorhe) [reply src/coreclr/vm/mono/mono_coreclr.cpp:500]: I can see how the comment makes it harder to follow.  The native code is not generated.  However, there is still a C# method that is generated.

The changes here to `mono_class_from_name` can be dropped.  You can keep the other change in this PR if you'd like.

File: src/coreclr/vm/mono/mono_coreclr.cpp (hunk 2 above)
Comment (joncham) [reply src/coreclr/vm/mono/mono_coreclr.cpp:508]: I agree that we should make exceptions (and exceptional cases) fatal.

File: src/coreclr/vm/mono/mono_coreclr.cpp (hunk 1 above)
Comment (alexey-zakharov) [reply src/coreclr/vm/mono/mono_coreclr.cpp:500]: I confirm that the #178 changes make exception to be added to the log.
it looks nice and clean like
```
TypeLoadException: Type 'UnityEngine.U2D.IK.IKManager2D' from assembly 'Unity.2D.IK.Runtime, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' is attempting to implement an inaccessible interface.
   at System.Reflection.RuntimeAssembly.GetType(QCallAssembly assembly, String name, Boolean throwOnError, Boolean ignoreCase, ObjectHandleOnStack type, ObjectHandleOnStack keepAlive, ObjectHandleOnStack assemblyLoadContext) in C:\build\output\Unity-Technologies\runtime\src\coreclr\System.Private.CoreLib\Microsoft.Interop.LibraryImportGenerator\Microsoft.Interop.LibraryImportGenerator\LibraryImports.g.cs:line 1221
   at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase) in C:\build\output\Unity-Technologies\runtime\src\coreclr\System.Private.CoreLib\src\System\Reflection\RuntimeAssembly.cs:line 214
   at Unity.CoreCLRHelpers.CoreCLRHost.class_from_name(IntPtr image, SByte* name_space, SByte* name, Boolean ignoreCase) in C:\build\output\Unity-Technologies\runtime\unity\unity-embed-host\CoreCLRHost.cs:line 92
   at Unity.CoreCLRHelpers.CoreCLRHost.class_from_name_native(IntPtr image, SByte* name_space, SByte* name, Boolean ignoreCase) in C:\build\output\Unity-Technologies\runtime\unity\unity-embed-host\UnityEmbedHost.Generator\UnityEmbedHost.Generator.CallbacksGenerator\GeneratedCoreCLRHost.gen.cs:line 60
UnityEngine_CoreModule_ProxyFunctionPointers:UnityEngine_StackTraceUtility_ExtractStringFromExceptionInternal_UnmanagedCallers(IntPtr, ManagedSpanWrapper*, ManagedSpanWrapper*, IntPtr)
UnityEngine.DebugLogHandler:Internal_LogException(Exception, Object)
UnityEngine.DebugLogHandler:LogException(Exception, Object)
UnityEngine.Logger:LogException(Exception, Object)
UnityEngine.Debug:LogException(Exception)
UnityEngine.<>c:<RegisterUECatcher>b__0_0(Object, UnhandledExceptionEventArgs) (at C:\UnitySrc\neutron.git\Runtime\Export\Scripting\UnhandledExceptionHandler.bindings.cs:46)
System.Reflection.RuntimeAssembly:GetType(QCallAssembly, String, Boolean, Boolean, ObjectHandleOnStack, ObjectHandleOnStack, ObjectHandleOnStack) (at C:\build\output\Unity-Technologies\runtime\src\coreclr\System.Private.CoreLib\Microsoft.Interop.LibraryImportGenerator\Microsoft.Interop.LibraryImportGenerator\LibraryImports.g.cs:1221)
System.Reflection.RuntimeAssembly:GetType(String, Boolean, Boolean) (at C:\build\output\Unity-Technologies\runtime\src\coreclr\System.Private.CoreLib\src\System\Reflection\RuntimeAssembly.cs:214)
Unity.CoreCLRHelpers.CoreCLRHost:class_from_name(IntPtr, SByte*, SByte*, Boolean) (at C:\build\output\Unity-Technologies\runtime\unity\unity-embed-host\CoreCLRHost.cs:92)
Unity.CoreCLRHelpers.CoreCLRHost:class_from_name_native(IntPtr, SByte*, SByte*, Boolean) (at C:\build\output\Unity-Technologies\runtime\unity\unity-embed-host\UnityEmbedHost.Generator\UnityEmbedHost.Generator.CallbacksGenerator\GeneratedCoreCLRHost.gen.cs:60)

(Filename: C:/build/output/Unity-Technologies/runtime/unity/unity-embed-host/CoreCLRHost.cs Line: 92)
```

thanks @mrvoorhe ! 
I don't think we need to land this PR - will close and delete branch

**Pull request closed without merging**